src/mscorlib/src/System/Environment.cs

```diff
@@ -536,23 +537,33 @@ private static void ValidateTarget(EnvironmentVariableTarget target)
 
         private static string GetEnvironmentVariableCore(string variable)
         {
-            StringBuilder sb = StringBuilderCache.Acquire(128); // A somewhat reasonable default size
-            int requiredSize = Win32Native.GetEnvironmentVariable(variable, sb, sb.Capacity);
+            Span<char> buffer = stackalloc char[128]; // A somewhat reasonable default size
+            return GetEnvironmentVariableCoreHelper(variable, buffer);
+        }
+
+        private static string GetEnvironmentVariableCoreHelper(string variable, Span<char> buffer)
+        {
+            int requiredSize = Win32Native.GetEnvironmentVariable(variable, buffer);
 
-            if (requiredSize == 0 && Marshal.GetLastWin32Error() == Win32Native.ERROR_ENVVAR_NOT_FOUND)
+            if (requiredSize > buffer.Length)
             {
-                StringBuilderCache.Release(sb);
-                return null;
+                char[] chars = ArrayPool<char>.Shared.Rent(requiredSize);
+                try
+                {
+                    return GetEnvironmentVariableCoreHelper(variable, chars);
+                }
+                finally
+                {
+                    ArrayPool<char>.Shared.Return(chars, clearArray: true);
+                }
             }
 
-            while (requiredSize > sb.Capacity)
+            if (requiredSize == 0)
             {
-                sb.Capacity = requiredSize;
-                sb.Length = 0;
-                requiredSize = Win32Native.GetEnvironmentVariable(variable, sb, sb.Capacity);
+                return Marshal.GetLastWin32Error() == Win32Native.ERROR_ENVVAR_NOT_FOUND ? null : string.Empty;
```

Comment:
Nit: I would move the check for error condition right after the PInvoke to make it easier to see that there is nothing modifying `Marshal.GetLastWin32Error()`.

Reply:
Done. I also changed the condition back to the way it was as the string ctor will take care of returning `string.Empty` for empty spans (when `requiredSize == 0` and the error isn’t “not found”).